client/clientservice/src/state_snapshot_service.cpp

```diff
@@ -22,27 +28,388 @@ using vmware::concord::client::statesnapshot::v1::StreamSnapshotResponse;
+                         bft::client::ByzantineSafeQuorum{}};
+
+  std::vector<std::unique_ptr<StateSnapshotReadAsOfResponse>> read_as_of_responses;
+  grpc::Status return_status;
+  auto callback =
+      [&read_as_of_responses, &return_status, &read_config](concord::client::concordclient::SendResult&& send_result) {
+        ResponseType<StateSnapshotReadAsOfResponse> reponse;
+        getResponseSetStatus(std::move(send_result),
+                             reponse,
+                             return_status,
+                             read_config.request.correlation_id,
+                             true,
+                             "concord.client.clientservice.state_snapshot_service.signedhash.callback");
+        read_as_of_responses =
+            std::get<std::vector<std::unique_ptr<StateSnapshotReadAsOfResponse>>>(std::move(reponse.response));
+      };
+
+  ConcordAssertNE(proto_request, nullptr);
+  ReconfigurationRequest rreq;
+  StateSnapshotReadAsOfRequest cmd;
+  cmd.snapshot_id = proto_request->snapshot_id();
+  cmd.participant_id = client_->getSubscriptionId();
+  for (int i = 0; i < proto_request->keys_size(); ++i) {
+    cmd.keys.push_back(proto_request->keys(i));
+  }
+  rreq.command = cmd;
+  bft::client::Msg message;
+  concord::messages::serialize(message, rreq);
+  client_->send(read_config, std::move(message), callback);
+  for (const auto& res : read_as_of_responses) {
+    if (return_status.ok()) {
+      switch (res->status) {
+        case SnapshotResponseStatus::InternalError:
+          return_status = grpc::Status(grpc::StatusCode::UNKNOWN, "Internal error");
+          break;
+        case SnapshotResponseStatus::SnapshotNonExistent:
+          return_status = grpc::Status(grpc::StatusCode::UNAVAILABLE, "Snapshot doesn't exist");
+          break;
+        case SnapshotResponseStatus::SnapshotPending:
+          return_status = grpc::Status(grpc::StatusCode::UNAVAILABLE, "Snapshot not ready");
+          break;
+        case SnapshotResponseStatus::Success: {
+          auto rv_size = res->values.size();
+          auto pr_size = proto_request->keys_size();
+          if (rv_size != static_cast<decltype(rv_size)>(pr_size)) {
+            return_status = grpc::Status(grpc::StatusCode::UNKNOWN, "Internal error");
+          } else {
+            bool set_values = (response->values_size() == 0);
+            size_t pos = 0;
+            for (const auto& v : res->values) {
+              if (set_values) {
+                auto* resp_val = response->add_values();
+                if (v.has_value()) {
+                  resp_val->set_value(v.value());
+                }
+              } else {
+                if (v.has_value()) {
+                  if (v.value() != response->values(pos).value()) {
+                    return_status =
+                        grpc::Status(grpc::StatusCode::UNKNOWN, "Got different values from different replicas");
```

Comment:
The whole for-loop is a bit magic to me given that the variable names are as generic as it gets. I understand that we loop over all responses and then check if we already have a response "at that position". What is the position? Is it the number of the kv-pairs? Also, how can you tell that if we already have a value at that position that the current response value must be from a different replica?

Reply:
Thanks, @robem, for the question. Here we get the rsi from the bft client. Each replica sends its response. This response contains a signature. The response is defined in the cmf file and @dartdart26 can explain the reason for the structure of the response.
In this loop, there are 2 parts. One part assigns the first value from the rsi. If the bool set_value is true, it means we have to set the values as the value size for response is 0. So this is the first time. So we set the values. After that set_value will be false as the values are already set. Then we iterate to check if the values set from the first entry of rsi are matching with the other rsis. rsi == Replica Specific Information.
I can see the reason for your confusion, the set_value means we have to set the values. So I am changing the name of this variable to have_to_set_value. This logic is based on the response from the reconfiguration request. I am adding more comments as well here.
Thanks for this comment.